### PR TITLE
Issue A - Missing Check on Quorum for the RoundChange Justification

### PR DESCRIPTION
**Pull request by GalRogozinski**
### Synopsis
The function getRoundChangeJustification in the qbft/prepare.go file does not check that the
set of constructed Prepare messages is of size quorum. This is a deviation from the [QBFT formal
verification code](https://github.com/ConsenSys/qbft-formal-spec-and-verification/blob/1630128e7f5468c08983d08064230422d9337805/dafny/spec/L1/node_auxiliary_functions.dfy#L673).

### Impact
The function returns a set of valid Prepare messages that is attached to a RoundChange message to
justify the round change by the operator. As specified in the QBFT code, the operator needs to check here
the size of the set. Not checking this can lead to sending around a RoundChange message that is not
accepted by the other operators. This in turn can lead to liveness issues, i.e. the honest operators send
around invalid RoundChange messages that are not accepted by other honest operators. Since the
proposer for a higher round requires quorum-many valid RoundChange messages, this can lead to a
state in which the operators do not find consensus and liveness is not reached.

### Fix
Add missing check and spectest

### CODE REVIEW

File: qbft/spectest/tests/create_message_spectest.go [hunk 1]
```diff
@@ -17,8 +17,11 @@ const (
 )
 
 type CreateMsgSpecTest struct {
-	Name                                             string
-	Value                                            [32]byte
+	Name string
+	//TODO rename to root
+	Value [32]byte
+	//TODO rename to value
```
Comment (GalRogozinski) [inline qbft/spectest/tests/create_message_spectest.go:23]: The rename affects other test
after the PR is accepted I will open issues for the rename and remove the TODO comment

File: qbft/spectest/tests/create_message_spectest.go [hunk 2]
```diff
@@ -28,29 +31,34 @@ type CreateMsgSpecTest struct {
...
 		t.Fail()
 	}
 
-	r, err := msg.GetRoot()
-	if err != nil {
-		lastErr = err
+	if err != nil && len(test.ExpectedError) != 0 {
+		require.EqualError(t, err, test.ExpectedError)
+		return
+	} else {
```
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply qbft/spectest/tests/create_message_spectest.go:51]: fixed

**Comment by GalRogozinski** [comment]
State Comparison will be added here after a seperate PR will add this ability to `create_message_spectest.go`

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: qbft/prepare.go [hunk 3]
```diff
@@ -70,6 +70,10 @@ func getRoundChangeJustification(state *State, config IConfig, prepareMsgContain
 			ret = append(ret, msg)
 		}
 	}
+
+	if !HasQuorum(state.Share, ret) {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply qbft/prepare.go:77]: We simply make sure the justification will be `nil` (later on the code makes it an empty slice IIRC).

It is worth to say that the audit folks admitted (and I agree) that in practice even w.o the check there is no clear attack vector.

Nonetheless, it is a good check to have and it exists in Dafny

File: qbft/prepare.go [hunk 4]
```diff
@@ -70,6 +70,10 @@ func getRoundChangeJustification(state *State, config IConfig, prepareMsgContain
 			ret = append(ret, msg)
 		}
 	}
+
```
[the remark this answers is not in the document]
Comment (GalRogozinski) [reply qbft/prepare.go:75]: done

**Review by y0sher — APPROVED** [review]

**Review by alonmuroch — APPROVED** [review]

**Comment by GalRogozinski** [comment]
@alonmuroch 
> we have some real world scenario to test this out with?
you mean doing tests on a testnet?

And yes I agree with you. Only if we allow nil proposal values (I don't think we do) maybe then the new check can somehow be important.

But better to be danfy-spec compliant and defensive.

**Pull request merged**